MCprep_addon/spawner/mcmodel.py

```diff
@@ -0,0 +1,256 @@
+            elements, textures = read_model(os.path.join(models_dir,parent_filepath)+".json")
+    
+    current_elements=obj_data.get("elements")
+    if current_elements is not None:
+        elements = current_elements # overwrites any elements from parents
+
+    current_textures=obj_data.get("textures")
+    if current_textures is not None:
+        if textures is None: textures=current_textures
+        else:
+            for img in current_textures:
+                textures[img]=current_textures[img]
+
+    conf.log("\nfile:"+str(model_filepath))
+    conf.log("parent:"+str(parent))
+    conf.log("elements:"+str(elements))
+    conf.log("textures:"+str(textures))
+
+    return elements, textures
+
+def add_model(model_filepath, obj_name="MinecraftModel"):
+    
+    mesh = bpy.data.meshes.new(obj_name)  # add a new mesh
+    obj = bpy.data.objects.new(obj_name, mesh)  # add a new object using the mesh
+
+    collection = bpy.context.collection
+    view_layer = bpy.context.view_layer
+    collection.objects.link(obj)  # put the object into the scene (link)
+    view_layer.objects.active = obj  # set as the active object in the scene
+    obj.select_set(True)  # select object
+
+    mesh = bpy.context.object.data
+    bm = bmesh.new()
+
+    mesh.uv_layers.new()
+    uv_layer = bm.loops.layers.uv.verify()
+    elements, textures = read_model(model_filepath)
+
+    materials=[]
+    for img in textures:
+        if img!="particle":
+            tex_pth=locate_image(textures,img,model_filepath)
+            mat = add_material(obj_name+"_"+img, tex_pth)
+            obj.data.materials.append(mat)
+            materials.append("#"+img)
+
+    if elements is None:
+        elements=[{'from':[0,0,0],'to':[0,0,0]}] # temp default elements
+
+    for e in elements:
+        rotation = e.get("rotation")
+        if rotation is None:
+            rotation={"angle": 0, "axis": "y", "origin": [8, 8, 8]} # rotation default
+        
+        element = add_element(e['from'],e['to'],rotation['origin'],rotation['axis'],rotation['angle']) 
+        verts=[bm.verts.new(v) for v in element[0]]  # add a new vert
+        uvs=[[1,1],[0,1],[0,0],[1,0]]
+        face_dir=["north","south","up","down","west","east"] # face directions defaults
+        faces = e.get("faces")
+        for i in range(len(element[2])):
+            f=element[2][i]
+            if faces:
+                d_face=faces.get(face_dir[i])
+                if d_face:
+                    face=bm.faces.new((verts[f[0]],verts[f[1]],verts[f[2]],verts[f[3]]))
+                    face.normal_update()
+
+                    uv_rot = d_face.get("rotation")
+                    if uv_rot is None:
+                        uv_rot = 0
+
+                    uv_idx = int(uv_rot/90)
+
+                    uv_cords = d_face.get("uv")
+                    if uv_cords is None:
+                        uv_cords=[0,0,16,16]
+                        
+                    uvs=[
```

Comment:
Could you add a comment here describing (one-liner or two liner at most, if possible) the logic behind these indices? e.g. why is it 2,0,0,2 and 1,1,3,3, and functionally what is this subtraction representing/doing?